OsmAnd/src/net/osmand/plus/server/endpoints/TileEndpoint.java

```diff
@@ -0,0 +1,95 @@
+package net.osmand.plus.server.endpoints;
+
+import android.graphics.Bitmap;
+import android.util.Log;
+import fi.iki.elonen.NanoHTTPD;
+import net.osmand.PlatformUtil;
+import net.osmand.data.RotatedTileBox;
+import net.osmand.plus.OsmandApplication;
+import net.osmand.plus.render.MapRenderRepositories;
+import net.osmand.plus.resources.AsyncLoadingThread;
+import net.osmand.plus.server.ApiEndpoint;
+import net.osmand.plus.server.OsmAndHttpServer;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+
+import static fi.iki.elonen.NanoHTTPD.newFixedLengthResponse;
+
+public class TileEndpoint implements ApiEndpoint {
+	private static final int RENDER_WAIT_THRESHOLD = 5000;
+	private static final Object lock = new Object();
+	private static final int TILE_SIZE_PX = 512;
+	Map<RotatedTileBox, Bitmap> hashMap = new HashMap<>();
+	Map<RotatedTileBox, Bitmap> map = Collections.synchronizedMap(hashMap);
+	OsmandApplication application;
+
+	public TileEndpoint(OsmandApplication application) {
+		this.application = application;
+	}
+
+	@Override
+	public NanoHTTPD.Response process(NanoHTTPD.IHTTPSession session) {
+		synchronized (lock) {
+			int zoom;
+			double lat;
+			double lon;
+			String fullUri = session.getUri().replace("/tile/", "");
```

Comment:
TileEndpoint shouldn't be aware what path is used
it could be "/tile/" or "/tilehd" or whatever, so url path probably should be passed as argument